src/stan/services/util/initialize.hpp

```diff
@@ -13,202 +13,211 @@
+      logger.info(e.what());
+      throw;
+    }
+    if (!boost::math::isfinite(log_prob)) {
+      logger.info("Rejecting initial value:");
+      logger.info("  Log probability evaluates to log(0),"
+                  " i.e. negative infinity.");
+      logger.info("  Stan can't start sampling from this"
+                  " initial value.");
+      continue;
+    }
+    std::stringstream log_prob_msg;
+    std::vector<double> gradient;
+    bool gradient_ok = true;
+    clock_t start_check = clock();
+    try {
+      log_prob = stan::model::log_prob_grad<true, true>
+                 (model, unconstrained, disc_vector,
+                  gradient, &log_prob_msg);
+    } catch (const std::exception& e) {
+      if (log_prob_msg.str().length() > 0)
+        logger.info(log_prob_msg);
+      logger.info(e.what());
+      throw;
+    }
+    clock_t end_check = clock();
+    double deltaT = static_cast<double>(end_check - start_check)
+                    / CLOCKS_PER_SEC;
+    if (log_prob_msg.str().length() > 0)
+      logger.info(log_prob_msg);
 
-            logger.info("Adjust your expectations accordingly!");
-            logger.info("");
-            logger.info("");
-          }
-          if (gradient_ok)
-            break;
-        }
+    for (size_t i = 0; i < gradient.size(); ++i) {
+      if (gradient_ok && !boost::math::isfinite(gradient[i])) {
+        logger.info("Rejecting initial value:");
+        logger.info("  Gradient evaluated at the initial value"
+                    " is not finite.");
+        logger.info("  Stan can't start sampling from this"
+                    " initial value.");
+        gradient_ok = false;
+      }
+    }
+    if (gradient_ok && print_timing) {
+      logger.info("");
+      std::stringstream msg1;
+      msg1 << "Gradient evaluation took " << deltaT << " seconds";
+      logger.info(msg1);
 
-        if (num_init_tries == MAX_INIT_TRIES) {
-          if (!is_fully_initialized && !init_zero) {
-            logger.info("");
-            std::stringstream msg;
-            msg << "Initialization between (-" << init_radius
-                << ", " << init_radius << ") failed after"
-                << " " << MAX_INIT_TRIES <<  " attempts. ";
-            logger.info(msg);
-            logger.info(" Try specifying initial values,"
-                        " reducing ranges of constrained values,"
-                        " or reparameterizing the model.");
-          }
-          throw std::domain_error("Initialization failed.");
-        }
+      std::stringstream msg2;
+      msg2 << "1000 transitions using 10 leapfrog steps"
+           << " per transition would take"
+           << " " << 1e4 * deltaT << " seconds.";
+      logger.info(msg2);
 
-        init_writer(unconstrained);
-        return unconstrained;
-      }
+      logger.info("Adjust your expectations accordingly!");
+      logger.info("");
+      logger.info("");
+    }
+    if (gradient_ok)
```

Comment:
This should return from here---there's no more work to do, then you won't need a confusing test in the next block.

```
if (gradient_ok) {
  init_writer(unconstrained);
  return unconstrained;
}
```

Then the rest of the logic simplifies.